# Creating a gold-standard dataset from pf8 and GenRE MEKONG data

The aim of this notebook is to create a gold-standard dataset for malaria genomics pipelines that predict drug resistance.

The following resources will be used:
1. [MalariaGEN Pf8](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)
2. [GenRE Mekong](https://www.malariagen.net/resource/29/)

## Rationale for selection of resources
Pipelines are usually tailored to a particular sequences technology, such as whole-genome or amplicon sequencing. While Pf8 is based on (selective) Whole Genome Sequencing (sWGS) data, GenRE Mekong utilised the SPotMalaria panel for amplicon sequencing (download available at [https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx).
Some GenRE Mekong samples have been re-analysed with sWGS for Pf8, thus providing data for the same sample from two different technologies and two independent resistance phenotype calls.
We will use the intersect between Pf8 and GenRE Mekong as the gold-standard dataset.

In Pf8, drug-resistance phenotypes are inferred based on criteria described in [this PDF](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf)

## Goal
THe notebook creates a csv file of sample IDs and metadata. The sample IDs can be used to retrieve the raw read data from public databases.  
This can be done manually or by using a tool such as [nf-core fetchngs](https://nf-co.re/fetchngs/1.12.0).


## Set up environment
All packages imported here must be installed locally

In [1]:
import pandas as pd
import requests
import re
from warnings import simplefilter 
simplefilter(action="ignore", category=pd.errors.PerformanceWarning)
pd.options.mode.copy_on_write = True

___custom ENA data helper module___
If this import fails, check that the ENA_data_helper.py script exists in the same directory as this Jupyter notebook. 

In [2]:
from ENA_data_helper import create_ena_data_frame, merge_ena_results_into_sample_data_genre_pf8

## Load Pf8 data
Obtain sample metadata and drug resisance inference data from Pf8.
Details can be found [here](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)

In [3]:
pf8_infer_resistance_df=pd.read_csv("https://pf8-release.cog.sanger.ac.uk/Pf8_inferred_resistance_status_classification.tsv", sep="\t")

In [4]:
pf8_samples_df=pd.read_csv("https://pf8-release.cog.sanger.ac.uk/metadata/Pf8_samples.txt", sep="\t")

## Merge Pf8 resistance and sample metadata
Also rename the sample column with lower case for consistency across data

In [5]:
pf8_samples_df.rename(columns={'Sample': 'sample'}, inplace=True)

In [6]:
pf8_df = pd.merge(pf8_infer_resistance_df, pf8_samples_df, on=["sample"], how="outer")

In [7]:
pf8_df

sample   Chloroquine Pyrimethamine   Sulfadoxine    Mefloquine  \
0       FP0008-C  Undetermined  Undetermined  Undetermined     Sensitive   
1       FP0009-C     Resistant     Resistant     Sensitive     Sensitive   
2      FP0010-CW  Undetermined     Resistant     Resistant     Sensitive   
3      FP0011-CW  Undetermined     Resistant  Undetermined     Sensitive   
4      FP0012-CW     Resistant     Resistant     Sensitive     Sensitive   
...          ...           ...           ...           ...           ...   
33320   SPT92268           NaN           NaN           NaN           NaN   
33321   SPT92269           NaN           NaN           NaN           NaN   
33322   SPT92270           NaN           NaN           NaN           NaN   
33323   SPT94772     Resistant     Resistant     Resistant  Undetermined   
33324   SPT94773     Resistant     Resistant     Resistant     Sensitive   

        Artemisinin   Piperaquine SP (uncomplicated)  SP (IPTp)         AS-MQ  \
0         Sensitive     Sensitive          Sensitive  Sensitive     Sensitive   
1         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   
2         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   
3         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   
4         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   
...             ...           ...                ...        ...           ...   
33320           NaN           NaN                NaN        NaN           NaN   
33321           NaN           NaN                NaN        NaN           NaN   
33322           NaN           NaN                NaN        NaN           NaN   
33323  Undetermined  Undetermined          Resistant  Sensitive  Undetermined   
33324     Sensitive     Sensitive          Resistant  Sensitive     Sensitive   

       ... Admin level 1 longitude    Year          ENA All samples same case  \
0      ...               -9.832345  2014.0   ERR1081237              FP0008-C   
1      ...               -9.832345  2014.0   ERR1081238              FP0009-C   
2      ...               -9.832345  2014.0   ERR2889621             FP0010-CW   
3      ...               -9.832345  2014.0   ERR2889624             FP0011-CW   
4      ...               -9.832345  2014.0   ERR2889627             FP0012-CW   
...    ...                     ...     ...          ...                   ...   
33320  ...                3.618846  2000.0  ERR10940681              SPT92268   
33321  ...                3.618846  2000.0  ERR11009733              SPT92269   
33322  ...                3.618846  2000.0  ERR11009737              SPT92270   
33323  ...              -16.401559  2017.0  ERR10789456              SPT94772   
33324  ...              -16.401559  2017.0  ERR10789457              SPT94773   

       Population  % callable  QC pass  Exclusion reason  Sample type  \
0            AF-W       82.48     True      Analysis_set         gDNA   
1            AF-W       88.95     True      Analysis_set         gDNA   
2            AF-W       87.01     True      Analysis_set         sWGA   
3            AF-W       86.95     True      Analysis_set         sWGA   
4            AF-W       89.86     True      Analysis_set         sWGA   
...           ...         ...      ...               ...          ...   
33320        AF-W        0.10    False      Low_coverage         sWGA   
33321        AF-W        0.12    False      Low_coverage         sWGA   
33322        AF-W        0.01    False      Low_coverage         sWGA   
33323        AF-W       64.96     True      Analysis_set         sWGA   
33324        AF-W       90.53     True      Analysis_set         sWGA   

      Sample was in Pf7  
0                  True  
1                  True  
2                  True  
3                  True  
4                  True  
...                 ...  
33320             False  
33321             False  
33322             False  
33323         

## Load GenRe Mekong data
Using this file from the GenRe Mekong webstie, release v1.4 14/05/24 (latest release):  

https://assets.super.so/abd63e6d-92df-4a9f-b19c-3a65d59b8a34/files/4accef63-0436-4893-a00a-6ea0e8a55249/Pf-GenRe-GRCv1.4-Full-PublicRelease-20240514.xlsx  
Tab: GRC

Sample IDs for samples that were re-analysed in Pf8 are the same in both projects, i.e. it is possible to merge GenRe Mekong and Pf8 data on sample ID.  

For more information on the data in this spreadsheet, consult the [GenRe Mekong data dictionary](https://github.com/GenRe-Mekong/Documents/blob/e331501b8e93752221ebd7ae72e5eddd3e6e924d/Pf-GenRe-GRCv1.4-DataDictionary.xlsx)

In [8]:
GenRe_df = pd.read_excel(
    "https://assets.super.so/abd63e6d-92df-4a9f-b19c-3a65d59b8a34/files/4accef63-0436-4893-a00a-6ea0e8a55249/Pf-GenRe-GRCv1.4-Full-PublicRelease-20240514.xlsx",
    sheet_name="GRC"
)

# make sample ID column name match Pf8
GenRe_df.rename(columns={'SampleId': 'sample'}, inplace=True)

## Merge GenRe Mekong and Pf8 data
Sample IDs for samples that were analysed in Pf8 are not changed, i.e. we can merge GenRe and Pf8 on sample ID.  
Retain only rows that have a sample ID that is found in GenRe and Pf8.

In [9]:
genre_pf8_merged_df = pd.merge(GenRe_df, pf8_df, on=["sample"], suffixes=('_GRMK','_pf8'), how="inner") 

## Remove GenRe Agena data
In the GenRe Mekong project, three techniques were used:
- Agena
- AmpSeqV1
- AmpSeqV2

These are recorded in the "Process" column of the GenRe Mekong data. Only AmpSeq data is relevant here because we are building a dataset for genotyping by sequencing. "Agena" refers to ["Agena Bioscience MassARRAY"](https://link.springer.com/protocol/10.1007/978-1-4939-6442-0_5), a technique based on PCR amplification of SNP regions that does not use sequencing and should thus not be used in this dataset.  

Count samples by process used in GenRe Mekong and "Sample type" in Pf8.  The Pf8 sample type, according to the [Pf8 data release README](https://pf8-release.cog.sanger.ac.uk/Pf8_README.txt), refers to
"Amplification technology used on the sample (MDA, gDNA or sWGA)", where "gDNA" means non-selective WGS, as opposed to selective sWGS. 

In [10]:
genre_pf8_merged_df.groupby(['Process','Sample type']).size().to_frame('count').reset_index()

Process Sample type  count
0     Agena        sWGA   3690
1  AmpSeqV1        gDNA     64
2  AmpSeqV1        sWGA    558
3  AmpSeqV2        sWGA   2730

-> 3690 samples in GenRe Mekong are based on Agena array, not on sequencing.  
Remove those samples.

In [11]:
genre_pf8_merged_seq_df = genre_pf8_merged_df[ genre_pf8_merged_df['Process'] != "Agena" ]

In [12]:
genre_pf8_merged_seq_df.groupby(['Process','Sample type']).size().to_frame('count').reset_index()

Process Sample type  count
0  AmpSeqV1        gDNA     64
1  AmpSeqV1        sWGA    558
2  AmpSeqV2        sWGA   2730

## Identify concordant drug-resistance calls and high-quality data
Both projects, GenRe Mekong and Pf8, provide drug-resistance phenotype predictions.  the final gold-standard dataset should contain only samples that have the same phenotype predictions in both projects and are of high quality.  


Identify columns that are found in both GenRe and Pf8

In [13]:
list(GenRe_df.columns.intersection( pf8_df.columns).values)

['sample',
 'Study',
 'Year',
 'Country',
 'Artemisinin',
 'Piperaquine',
 'Mefloquine',
 'Chloroquine',
 'Pyrimethamine',
 'Sulfadoxine',
 'DHA-PPQ',
 'AS-MQ']

-> all except 'sample', 'Study', 'Year' and 'Country'are drug-resistance phenotypes.  
Create a list of field names to compare.

In [14]:
phenotypes = [
 'Artemisinin',
 'Piperaquine',
 'Mefloquine',
 'Chloroquine',
 'Pyrimethamine',
 'Sulfadoxine',
 'DHA-PPQ',
 'AS-MQ']

Get counts for concordant vs. discordant phenotype calls in Pf8 and GenRe Mekong

In [15]:
discordant_counts = []
for p in phenotypes:
    discordant_counts.append(
        {
            'drug': p, 
            'discordant_count': len(
                genre_pf8_merged_seq_df[
                    genre_pf8_merged_seq_df[ p + '_GRMK'] != genre_pf8_merged_seq_df[ p + '_pf8']
                ]
            ),
            'concordant_count': len(
                genre_pf8_merged_seq_df[
                    genre_pf8_merged_seq_df[ p + '_GRMK'] == genre_pf8_merged_seq_df[ p + '_pf8']
                ]
            )
    })
pd.DataFrame(discordant_counts)


drug  discordant_count  concordant_count
0    Artemisinin               832              2520
1    Piperaquine              1807              1545
2     Mefloquine              1853              1499
3    Chloroquine               413              2939
4  Pyrimethamine               480              2872
5    Sulfadoxine               478              2874
6        DHA-PPQ              1784              1568
7          AS-MQ              1834              1518

-> the data shows that there are discordant as well as concordant calls for all drug resistance phenotypes, thus it is important to filter for samples that are concordant.

Filter for rows that are concordant on all 8 phenotypes

In [16]:
genre_pf8_conc_df = genre_pf8_merged_seq_df
for p in phenotypes:
    cols = [ p + '_GRMK', p + '_pf8' ]
    genre_pf8_conc_df = genre_pf8_conc_df[
        ( genre_pf8_conc_df[cols].notnull().all(axis='columns') ) &
        ( genre_pf8_conc_df[cols[0]] == genre_pf8_conc_df[cols[1]] )
    ]

Quality filter: percent of callable bases > 85 and 'QC pass' is True

In [17]:
genre_pf8_conc_df = genre_pf8_conc_df[ 
    (genre_pf8_conc_df['% callable'] >= 85 ) &
    (genre_pf8_conc_df['QC pass'] == True )
]

Number of rows after filtering

In [18]:
len(genre_pf8_conc_df)

968

Reduce the dataset to a smaller number of relevant columns.

In [19]:
keep_cols = [
        'sample', 'Artemisinin_GRMK', 'Piperaquine_GRMK',
       'Mefloquine_GRMK', 'Chloroquine_GRMK', 'Pyrimethamine_GRMK',
       'Sulfadoxine_GRMK', 'DHA-PPQ_GRMK', 'AS-MQ_GRMK', 'Species', 
       'Chloroquine_pf8', 'Pyrimethamine_pf8',
       'Sulfadoxine_pf8', 'Mefloquine_pf8', 'Artemisinin_pf8',
       'Piperaquine_pf8', 'AS-MQ_pf8',
       'DHA-PPQ_pf8', 'ENA', '% callable'
]

genre_pf8_conc_s_df = genre_pf8_conc_df[keep_cols]

Combine the phenotype calls into single column each (they are all concordant due to the filter used)

In [20]:
for p in phenotypes:
    # just to be safe
    if not genre_pf8_conc_s_df[p + '_GRMK'].equals(genre_pf8_conc_s_df[p + '_pf8']):
        raise Exception('phenotypes are not concordant but they should be if filtered correctly')
    genre_pf8_conc_s_df = genre_pf8_conc_s_df.rename(columns={p + '_GRMK': p}, errors="raise")
    genre_pf8_conc_s_df = genre_pf8_conc_s_df.drop([ p + '_pf8' ], axis=1)

Overview of final data

In [21]:
genre_pf8_conc_s_df

sample Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \
3730  RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   
3731  RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   
3733  RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   
3734  RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   
3735  RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   
...        ...         ...         ...        ...         ...           ...   
6975  RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   
6976  RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   
6987  RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   
7021  RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   
7025  RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   

     Sulfadoxine    DHA-PPQ      AS-MQ Species         ENA  % callable  
3730   Resistant  Sensitive  Sensitive      Pf  ERR3409305       89.14  
3731   Resistant  Resistant  Sensitive      Pf  ERR3409312       85.83  
3733   Resistant  Resistant  Sensitive      Pf  ERR3409373       87.23  
3734   Resistant  Resistant  Sensitive      Pf  ERR3409314       87.04  
3735   Resistant  Sensitive  Sensitive      Pf  ERR3409311       88.65  
...          ...        ...        ...     ...         ...         ...  
6975   Resistant  Sensitive  Sensitive      Pf  ERR7221085       87.92  
6976   Sensitive  Sensitive  Sensitive      Pf  ERR7221086       87.30  
6987   Resistant  Sensitive  Sensitive      Pf  ERR7221102       89.35  
7021   Resistant  Sensitive  Sensitive      Pf  ERR7221141       86.03  
7025   Resistant  Sensitive  Sensitive      Pf  ERR7221132       89.21  

[968 rows x 12 columns]

## Add ENA accessions and download links
Using functions from the custom ENA data helper tool (file ENA_data_helper.py in the working directory), perform a search on ENA by sample ID and retrieve data for run accessions (unique accession for 1 sample on 1 run) and FTP download URLs for FASTQ files (read1/read2).  
This step performs a live search on ENA and may not work if your internet connection is very slow or unstable.

The first function performs a search for ENA data by sample IDs in sample data frame

In [22]:
genre_pf8_conc_s_ena_result = create_ena_data_frame(genre_pf8_conc_s_df)

Second function merges the ENA search results (which contain more than one result per sample) back into the original sample data frame where sample ID is unique. Columns are added for the run accessions for all threee panels in GenRe as well as the WGS data for Pf8. 

In [23]:
genre_pf8_conc_s_ena_df = merge_ena_results_into_sample_data_genre_pf8(sample_data=genre_pf8_conc_s_df, ena_result=genre_pf8_conc_s_ena_result)

In [24]:
genre_pf8_conc_s_ena_df

Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \
sample                                                                  
RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   
RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   
RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   
RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   
RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   
...              ...         ...        ...         ...           ...   
RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   
RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   
RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   
RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   
RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   

         Sulfadoxine    DHA-PPQ      AS-MQ Species         ENA  ...  \
sample                                                          ...   
RCN12025   Resistant  Sensitive  Sensitive      Pf  ERR3409305  ...   
RCN12026   Resistant  Resistant  Sensitive      Pf  ERR3409312  ...   
RCN12028   Resistant  Resistant  Sensitive      Pf  ERR3409373  ...   
RCN12031   Resistant  Resistant  Sensitive      Pf  ERR3409314  ...   
RCN12032   Resistant  Sensitive  Sensitive      Pf  ERR3409311  ...   
...              ...        ...        ...     ...         ...  ...   
RCN26767   Resistant  Sensitive  Sensitive      Pf  ERR7221085  ...   
RCN26775   Sensitive  Sensitive  Sensitive      Pf  ERR7221086  ...   
RCN26820   Resistant  Sensitive  Sensitive      Pf  ERR7221102  ...   
RCN26932   Resistant  Sensitive  Sensitive      Pf  ERR7221141  ...   
RCN26938   Resistant  Sensitive  Sensitive      Pf  ERR7221132  ...   

          ENA_acc_GenRe_GRC2 ENA_acc_GenRe_SPEC  \
sample                                            
RCN12025         ERR14388604        ERR14388605   
RCN12026         ERR14388607        ERR14388608   
RCN12028         ERR14388613        ERR14388614   
RCN12031         ERR14388622        ERR14388623   
RCN12032         ERR14388625        ERR14388626   
...                      ...                ...   
RCN26767         ERR14397547        ERR14397548   
RCN26775         ERR14397571        ERR14397572   
RCN26820         ERR14397664        ERR14397665   
RCN26932         ERR14397976        ERR14397977   
RCN26938         ERR14397991        ERR14397992   

                                 GenRe_GRC1_ENA_FASTQ_FTP_1  \
sample                                                        
RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   
RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   
RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   
RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   
RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   
...                                                     ...   
RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   
RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   
RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   
RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   
RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   

                                 GenRe_GRC2_ENA_FASTQ_FTP_1  \
sample                                                        
RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   
RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   
RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   
RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   
RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   
...                                                     ...   
RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   
RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   
RCN26820  ftp://ftp.s

Keep only samples that have downloadable data for all three AmpSeq panels and WGS

In [25]:
genre_pf8_conc_s_ena2_df = genre_pf8_conc_s_ena_df[
    genre_pf8_conc_s_ena_df[['GenRe_GRC1_ENA_FASTQ_FTP_1',
       'GenRe_GRC2_ENA_FASTQ_FTP_1', 'GenRe_SPEC_ENA_FASTQ_FTP_1',
       'Pf8_ENA_FASTQ_FTP_1', 'GenRe_GRC1_ENA_FASTQ_FTP_2',
       'GenRe_GRC2_ENA_FASTQ_FTP_2', 'GenRe_SPEC_ENA_FASTQ_FTP_2',
       'Pf8_ENA_FASTQ_FTP_2']].notnull().all(1)
]

number of samples with and without the filter for FTP links

In [26]:
print(f'n samples before FTP link filter: {len(genre_pf8_conc_s_ena_df)}')
print(f'n samples with all FASTQ files available via FTP: {len(genre_pf8_conc_s_ena2_df)}')

n samples before FTP link filter: 968
n samples with all FASTQ files available via FTP: 907


### Save results: concordant phenotype data
Save to csv file
```Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv```  

In summary: this table of 907 rows represents samples that have been sequenced by amplicon sequencing for GenRe Mekong and WGS for Pf8 and where the  sample is marked as high-quality in Pf8 and all phenotye calls (inferred from genotypes) are fully conconrdant between Pf8 and GenRe Mekong.  
The samples have also been filtered to include only those that have FTP download links for all GenRe AmpSeq panels and the Pf8 WGS data.

In [27]:
genre_pf8_conc_s_ena2_df.to_csv('Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv', index=True)

***

## Add genotypes
In the above steps, concordance was defined on high-level drug resistance phenotype calls.  
Adding the underlying genotype calls will provide a lower-level truth dataset that can be used to validate genotype calls of a novel pipeline.  
___Note___ that "genotype" calls are provided at amino acid level - not nucelotide level - in both studies. 

### Genes and haplotypes
For the purpose of this gold-standard dataset, the intention is to identify samples that are fully concordant on their haplotype calls for important genes between Pf8 and GenRe Mekong, i.e. two different sequencing techinques and two pipelines have resulted in the same haplotype call.  

This dataset will focus on 8 key drug resistance genes where drug resistance is determined by non-synonymous point mutations:
- dhfr
- crt
- dhps
- kelch13
- mdr1
- mdr2
- fd
- arps10

This list of genes and "core" amino acid positions was compiled according to Table SM1 (Markers associated with drug resistance for P. falciparum) in the [SpotMalaria Platform Technical Notes and Methods](https://ngs.sanger.ac.uk/production/malaria/Resource/29/20200705-GenRe-04a-SpotMalaria-0.39.pdf). For Pf8, details about drug resistance classification can be found in in [this document](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf).  

This results in the following list of drug-resistance haplotypes to be considered for the gold- standard dataset:

| Gene | Amino Acid position |
|---|---|
| dhfr | 51, 59, 108, 164 |
| crt | 72, ~~73~~\*, 74, 75, 76, 326, 356 |
| dhps | 436, 437, 540, 581, 613 |
| kelch13 | 349-726 |
| mdr1 | 86, 184, 1246 |
| mdr2 | 484 |
| fd | 193 |
| arps10 | 127, 128 |
\* Position crt:73 is listed as a core position in the reference table but it is not provided in GenRe Mekong data and it is always WT in Pf8 data, hence this position will be ignored in the following code.  

For more details on gene haplotypes in P. falciparum see the [Pf-HaploAtlas](https://www.malariagen.net/article/introducing-pf-haploatlas-a-new-app-to-track-malaria-parasite-mutations/). 
       

### Load Pf8 genotype calls from data release
The GenRe data loaded earlier does contain genotype data already.

In [28]:
pf8_gt_df = pd.read_csv("https://pf8-release.cog.sanger.ac.uk/Pf8_drug_resistance_marker_genotypes.tsv", sep="\t")

### Merge Pf8 genotypes 
Merge into into existing dataframe  
```genre_pf8_merged_seq_df```  
created above, which is the merged data for Pf8 and GenRe Mekong, with array data removed. 


In [29]:
pf8_gt_df.rename(columns={'Sample': 'sample'}, inplace=True)

genre_pf8_merged_seq_gt_df = pd.merge(
    genre_pf8_merged_seq_df, 
    pf8_gt_df, 
    on=["sample"], 
    suffixes=('_GRMK','_pf8'), 
    how="inner") 

The dataframe now contains Pf8 and GenRe Mekong genotypes for all samples that have been sequenced (excluding Array data) in both projects and can be linked via sample ID.

### Create new dataframe with amino acid genotype calls
Create a list of gene/aa-positions from the above table and use it to comile a new data frame with amino acid calls from Pf8 and GenRe Mekong for these core positions. The kelch13 list needs to be built separately because the data is provided in separate column in GenRe Mekong and as a sigle column of comma-separated lists in Pf8.


The new columns will have a unified naming scheme:  
"core_mutation_{GENE}_{AAPos}_\[pf8/GMRK\]"

The majority of genotypes can be extracted from the Pf8 and GenRe data by a simple pattern.  
The respective pattern of genotype calls by gene and AA position is as follows:  
__Pf8__: {GENE}\_{AAPos} (example "crt_75\[N\]")  
__GenRe Mekong__: Pf{GENE}:{AAPos} (example: "PfCRT:75")  

Create a set of new column names and respective Pf8 and GenRe column names accordingly.  
There are ___two excpetions___: 
1. kelch13: in both datasets, kelch13 data is provided as a column of lists of AA positions that need to be converted separately
2. arps10: provided as AA pos 127-128 in Pf8 and just pos 127 in GenRe. Needs to be converted separately


In [30]:
genes_pos = {
    'dhfr': [51, 59, 108, 164],
    'crt': [72, 74, 75, 76, 326, 356], # omitting pos 73, see above for explanation
    'dhps': [436, 437, 540, 581, 613],
    'mdr1': [86, 184, 1246],
    'mdr2': [484],
    'fd': [193]
}

gt_cols = {}
for gene, pos_list in genes_pos.items():
    gt_cols[ gene ] = {}
    for pos in pos_list:
        gt_cols[ gene ][ pos ] = {
            'pf8_pattern': gene + '_' + str(pos) + '\[[A-Z]\]',
            'GRMK_name': 'Pf' + gene.upper() + ':' + str(pos),
            'new_col_basename': 'core_mutation_' + gene + '_' + str(pos)
        }

### Add new genotype columns 
Add new genotype columns for the core mutation positions to dataframe   
```genre_pf8_merged_seq_gt_df```

In [31]:
for gene in gt_cols:
    for pos in gt_cols[ gene ]:
        
        # genotype from Pf8
        pf8_pattern = gt_cols[gene][pos]['pf8_pattern']
        new_col_pf8 = gt_cols[gene][pos]['new_col_basename'] + '_pf8'
        pf8_col = [ c for c in genre_pf8_merged_seq_gt_df if re.match(pf8_pattern, c) ]
        if not pf8_col:
            raise ValueError(f'could not find Pf8 column starting with "{pf8_pattern}"')
        elif len(pf8_col) > 1:
            raise ValueError(f'found more than one Pf8 column starting with "{pf8_pattern}":{", ".join(pf8_col)}')
        else:
            genre_pf8_merged_seq_gt_df[new_col_pf8] = genre_pf8_merged_seq_gt_df[pf8_col[0]].copy()
            
        # genotype from GenRe
        genre_pattern = gt_cols[gene][pos]['GRMK_name']
        new_col_genre = gt_cols[gene][pos]['new_col_basename'] + '_GRMK'
        if genre_pattern in genre_pf8_merged_seq_gt_df:
            genre_pf8_merged_seq_gt_df[new_col_genre] = genre_pf8_merged_seq_gt_df[genre_pattern].copy()
        else:
            raise ValueError(f'could not find GenRe column "{genre_pattern}"')
                                                                         

Add columns for arps10: use only position arsp10:127, which is provided as a column inf GenRe and needs to be extracted from column 'arps10_127-128\[VD\]' in Pf8.

In [32]:
arps10_basename = 'core_mutation_arps10_127'
genre_pf8_merged_seq_gt_df[ arps10_basename + '_GRMK'] = genre_pf8_merged_seq_gt_df['PfARPS10:127'].copy()
genre_pf8_merged_seq_gt_df[ arps10_basename + '_pf8'] = genre_pf8_merged_seq_gt_df['arps10_127-128[VD]'].astype(str).str[0]

### Add columns for kelch13
Add columns for all kelch13 AA positions for which data exists in at least one sample in both projects.  
__Pf8__:  
    _column_: 'kelch13_349-726_ns_changes'  
    _content_: comma separated list of haplotypes at Kelch13 positions 349-726. Each haplotype contains one or more non-synonymous variations, separateed by "/". Homozygous mutations are shown in upper case and heterozygous in lower case.   Details are [here (heading "Pf8_drug_resistance_marker_genotypes.tsv"](https://pf8-release.cog.sanger.ac.uk/Pf8_README.txt)  
    
__GenRe__:  
    _column_: 'Pfkelch13'  
    _content_: Comma-separated list of amino acid mutations. ___NOTE___ this differs from Pf8. In GenRe, the comma delimits mutations in the same haplotype, while in Pf8 (see above) a comma delimts haplotypes. There is no disctiontion between homo- and heterozygous mutations.  For details see the [GenRe Mekong data dictionary](https://github.com/GenRe-Mekong/Documents/blob/e331501b8e93752221ebd7ae72e5eddd3e6e924d/Pf-GenRe-GRCv1.4-DataDictionary.xlsx)

Show the unique contents of the respective kelch13 columns in Pf8 and GenRe Mekong: 

### Create individual kelch13 mutation columns
For each individual mutation found in any sample in any project, create a column with a name following this schema:  
```core_mutation_kelch13:{MUTATION}_{pf8/GRMK}```  
MUTATION is of the format ```{WT allele}{Pos}{mutant allele}```   
The column contains boolean values where True indicates that this mutation is present in a given sample. Due to the differences in data provided between the two projects, no resolution at haplotype level is provided and no information about zygosity is recorded. Consequently, the "WT" allele recorded in some samples in GenRe Mekong is ignored. 

In [33]:
kelch13_mutations = set()

# Pf8
for mut_list in genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].unique():
    mut_list = str(mut_list)
    if mut_list and re.search(r'[1-9]', mut_list):
        mut_list = mut_list.replace('/',',') # treat all mutations individually, regardless of haplotype grouping
        mut_list = mut_list.replace('*','')
        for mut in mut_list.split(','):
            kelch13_mutations.add(mut.upper())
            
# GenRe Mekong
for mut_list in genre_pf8_merged_seq_gt_df['Pfkelch13'].unique():
    mut_list = str(mut_list)
    if mut_list and re.search(r'[1-9]', mut_list):
        for mut in mut_list.split(','):
            if mut != 'WT':
                kelch13_mutations.add(mut.upper())

# add the new columns
genre_pf8_merged_seq_gt_df = genre_pf8_merged_seq_gt_df.copy()
for mut in kelch13_mutations:
    base_name = 'core_mutation_kelch13:' + mut
    genre_pf8_merged_seq_gt_df[base_name + '_pf8'] = genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].str.contains(mut, case=False)
    genre_pf8_merged_seq_gt_df[base_name + '_GRMK'] = genre_pf8_merged_seq_gt_df['Pfkelch13'].str.contains(mut, case=False)
    
    

### Identify concordant samples
Filter for samples that are fully concordant on all core mutation genotype calls, i.e. all columns added with the "core_mutation_" prefix above.  

In [34]:
def is_concordant(row, *core_mut_col_basenames):
    """
    Return True if all 'core_mutation' columns match between 
    Pf8 and GenRe Mekong
    """
    for col in core_mut_col_basenames:
        if row[col + '_pf8'] != row[col + '_GRMK']:
            return False
    return True

regex = re.compile(r'^(core_mutation_.+)_(pf8|GRMK)$')
core_mut_col_basenames = set(
    [ regex.sub(r'\1', x) for x in genre_pf8_merged_seq_gt_df.columns.values if regex.match(x) ]
)
genre_pf8_merged_seq_gt_df = genre_pf8_merged_seq_gt_df.copy()
genre_pf8_merged_seq_gt_df['fully_concordant_gt'] = genre_pf8_merged_seq_gt_df.apply(is_concordant, axis=1, args=core_mut_col_basenames)


Count fully concordant (on all genotypes reported) vs. not fully concordant samples across the two projects:

In [35]:
genre_pf8_merged_seq_gt_df.groupby( ['fully_concordant_gt']).size().to_frame('count').reset_index()

fully_concordant_gt  count
0                False   1572
1                 True   1394

47% of samples are fully concordant on all drug-reistance "core mutation" genotypes between Pf8 (WGS) and GenRe Mekong (AmpSeq).

### Create a filtered dataset containing only fully concordant samples
In addition, apply filters on Pf8 quality, as done above for phenotypes (%callable >=85 and QC pass)

In [36]:
genre_pf8_merged_seq_gt_concord_df = genre_pf8_merged_seq_gt_df[
    (genre_pf8_merged_seq_gt_df['fully_concordant_gt'] == True ) &
    (genre_pf8_merged_seq_gt_df['% callable'] >= 85 ) &
    (genre_pf8_merged_seq_gt_df['QC pass'] == True )
]

In [37]:
len(genre_pf8_merged_seq_gt_concord_df)

976

There are 976 high-quality samples where all reported drug-resitance genotype positions are fully concordant between Pf8 and GenRe Mekong.

Keep only relevant columns. As all samples are fully concordant, keep only one column per core mutation and drop the pf8/GRMK suffix.

In [38]:
keep_cols = ['sample', 'Source', 'Process', 
    'TimePoint', 'Country_GRMK', 'Species',
    'fully_concordant_gt']

core_mut_cols = [x for x in genre_pf8_merged_seq_gt_concord_df.columns.values if x.startswith('core_mutation')]
keep_cols.extend( core_mut_cols )
genre_pf8_merged_seq_gt_concord_small_df = genre_pf8_merged_seq_gt_concord_df[keep_cols]
genre_pf8_merged_seq_gt_concord_small_df.rename({'Country_GRMK':'Country'})

regex = re.compile(r'^(core_mutation_.+)_(pf8|GRMK)$')
for col in core_mut_cols:
    col_basename = regex.sub(r'\1', col)
    genre_pf8_merged_seq_gt_concord_small_df[col_basename] = genre_pf8_merged_seq_gt_concord_small_df[col].copy()
    genre_pf8_merged_seq_gt_concord_small_df.drop(col, axis=1, inplace=True)

## Add ENA accessions and download links
Using functions from the custom ENA data helper tool (file ENA_data_helper.py in the working directory), perform a search on ENA by sample ID and retrieve data for run accessions (unique accession for 1 sample on 1 run) and FTP download URLs for FASTQ files (read1/read2).  
This step performs a live search on ENA and may not work if your internet connection is very slow or unstable.

The first function performs a search for ENA data by sample IDs in sample data frame

In [39]:
genre_pf8_merged_seq_gt_concord_small_ena_result = create_ena_data_frame(genre_pf8_merged_seq_gt_concord_small_df)

Second function merges the ENA search results (which contain more than one result per sample) back into the original sample data frame where sample ID is unique. Columns are added for the run accessions for all threee panels in GenRe as well as the WGS data for Pf8.

In [40]:
genre_pf8_merged_seq_gt_concord_small_ena_df = merge_ena_results_into_sample_data_genre_pf8(sample_data=genre_pf8_merged_seq_gt_concord_small_df, ena_result=genre_pf8_merged_seq_gt_concord_small_ena_result)

In [41]:
genre_pf8_merged_seq_gt_concord_small_ena_df

Source   Process TimePoint Country_GRMK Species  \
sample                                                            
RCN01775  GenRe-Mekong  AmpSeqV2         -           VN      Pf   
RCN01776  GenRe-Mekong  AmpSeqV2         -           VN      Pf   
RCN01777  GenRe-Mekong  AmpSeqV2         -           VN      Pf   
RCN01778  GenRe-Mekong  AmpSeqV2         -           VN      Pf   
RCN01779  GenRe-Mekong  AmpSeqV2         -           VN      Pf   
...                ...       ...       ...          ...     ...   
RCN26767  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   
RCN26782  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   
RCN26820  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   
RCN26852  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   
RCN26938  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   

          fully_concordant_gt core_mutation_dhfr_51 core_mutation_dhfr_59  \
sample                                                                      
RCN01775                 True                     I                     R   
RCN01776                 True                     I                     R   
RCN01777                 True                     I                     R   
RCN01778                 True                     I                     R   
RCN01779                 True                     I                     R   
...                       ...                   ...                   ...   
RCN26767                 True                     I                     R   
RCN26782                 True                     I                     R   
RCN26820                 True                     I                     R   
RCN26852                 True                     I                     R   
RCN26938                 True                     I                     R   

         core_mutation_dhfr_108 core_mutation_dhfr_164  ...  \
sample                                                  ...   
RCN01775                      N                      L  ...   
RCN01776                      N                      L  ...   
RCN01777                      N                      L  ...   
RCN01778                      N                      I  ...   
RCN01779                      N                      L  ...   
...                         ...                    ...  ...   
RCN26767                      N                      I  ...   
RCN26782                      N                      I  ...   
RCN26820                      N                      I  ...   
RCN26852                      N                      I  ...   
RCN26938                      N                      I  ...   

         ENA_acc_GenRe_GRC2 ENA_acc_GenRe_SPEC  \
sample                                           
RCN01775               None               None   
RCN01776               None               None   
RCN01777               None               None   
RCN01778               None               None   
RCN01779               None               None   
...                     ...                ...   
RCN26767        ERR14397547        ERR14397548   
RCN26782        ERR14397589        ERR14397590   
RCN26820        ERR14397664        ERR14397665   
RCN26852        ERR14397748        ERR14397749   
RCN26938        ERR14397991        ERR14397992   

                                 GenRe_GRC1_ENA_FASTQ_FTP_1  \
sample                                                        
RCN01775                                               None   
RCN01776                                               None   
RCN01777                                               None   
RCN01778                                               None   
RCN01779                                               None   
...                                                     ...   
RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   
RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/088/...   
RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fa

Keep only samples that have downloadable data for all three AmpSeq panels and WGS

In [42]:
genre_pf8_merged_seq_gt_concord_small_ena2_df = genre_pf8_merged_seq_gt_concord_small_ena_df[
    genre_pf8_merged_seq_gt_concord_small_ena_df[['GenRe_GRC1_ENA_FASTQ_FTP_1',
       'GenRe_GRC2_ENA_FASTQ_FTP_1', 'GenRe_SPEC_ENA_FASTQ_FTP_1',
       'Pf8_ENA_FASTQ_FTP_1', 'GenRe_GRC1_ENA_FASTQ_FTP_2',
       'GenRe_GRC2_ENA_FASTQ_FTP_2', 'GenRe_SPEC_ENA_FASTQ_FTP_2',
       'Pf8_ENA_FASTQ_FTP_2']].notnull().all(1)
]

number of samples with and without the filter for FTP links

In [43]:
print(f'n samples before FTP link filter: {len(genre_pf8_merged_seq_gt_concord_small_ena_df)}')
print(f'n samples with all FASTQ files available via FTP: {len(genre_pf8_merged_seq_gt_concord_small_ena2_df)}')

n samples before FTP link filter: 976
n samples with all FASTQ files available via FTP: 873


### Save file of fully concordant samples


In [44]:
genre_pf8_merged_seq_gt_concord_small_ena2_df.to_csv('Pf8-GenReMekong_fully_concordant_genotype_high_quality_samples.csv', index=True)

***

## Distinct genotype patterns and representative samples
Final dataset is based on the fully concordant genotypes dataset, but will only contain one representative sample for every distinct pattern of haplotypes.   
This dataset is based on the previous dataset, which means only samples with downloadable FASTQ files are taken into consideration here.

First, get a count of different patterns.

In [45]:
core_mut_cols = [x for x in genre_pf8_merged_seq_gt_concord_small_ena2_df.columns.values if x.startswith('core_mutation')]
genre_pf8_merged_seq_gt_concord_small_ena2_df.groupby( core_mut_cols).size().to_frame('count').reset_index()

core_mutation_dhfr_51 core_mutation_dhfr_59 core_mutation_dhfr_108  \
0                      I                     R                      N   
1                      I                     R                      N   
2                      I                     R                      N   
3                      I                     R                      N   
4                      I                     R                      N   
5                      I                     R                      N   
6                      I                     R                      N   
7                      I                     R                      N   
8                      I                     R                      N   
9                      I                     R                      N   
10                     I                     R                      N   
11                     I                     R                      N   
12                     I                     R                      N   
13                     I                     R                      N   
14                     I                     R                      N   
15                     I                     R                      N   
16                     I                     R                      N   
17                     I                     R                      N   
18                     I                     R                      N   
19                     I                     R                      N   
20                     I                     R                      N   
21                     I                     R                      N   
22                     I                     R                      N   
23                     I                     R                      N   
24                     I                     R                      N   
25                     I                     R                      N   
26                     I                     R                      N   
27                     I                     R                      N   
28                     I                     R                      N   
29                     I                     R                      N   
30                     I                     R                      N   
31                     I                     R                      N   
32                     I                     R                      N   
33                     I                     R                      N   

   core_mutation_dhfr_164 core_mutation_crt_72 core_mutation_crt_74  \
0                       I                    C                    I   
1                       I                    C                    I   
2                       I                    C                    I   
3                       I                    C                    I   
4                       I                    C                    I   
5                       I                    C                    I   
6                       I                    C                    I   
7                       I                    C                    I   
8                       I                    C                    I   
9                       I                    C                    I   
10                      I                    C                    I   
11                      I                    C                    I   
12                      I                    C                    I   
13                      I                    C                    I   
14                      I                    C                    I   
15                      I                    C                    I   
16                      I                    C                    I   
17                    I,L                    C                    I   
18                    I,L                    C                    I   
19                      L  

There are 34 distinct patterns of genotypes.
Extract a list of sample IDs of representative samples for each pattern.

In [47]:
genre_pf8_merged_seq_gt_concord_patterns_df = genre_pf8_merged_seq_gt_concord_small_ena2_df.reset_index().groupby( core_mut_cols)['sample'].min().reset_index().set_index('sample')

# move the samples column to first
cols = list(genre_pf8_merged_seq_gt_concord_patterns_df.columns)
genre_pf8_merged_seq_gt_concord_patterns_df = genre_pf8_merged_seq_gt_concord_patterns_df[cols]

# add the ENA accession and download link columns back in
genre_pf8_merged_seq_gt_concord_patterns_final_df = pd.merge(
    genre_pf8_merged_seq_gt_concord_patterns_df,
    genre_pf8_merged_seq_gt_concord_small_ena2_df.filter(regex='ENA_'),
    left_index=True,
    right_index=True
)
genre_pf8_merged_seq_gt_concord_patterns_final_df

core_mutation_dhfr_51 core_mutation_dhfr_59 core_mutation_dhfr_108  \
sample                                                                        
RCN13568                     I                     R                      N   
RCN13560                     I                     R                      N   
RCN15107                     I                     R                      N   
RCN14580                     I                     R                      N   
RCN13532                     I                     R                      N   
RCN13481                     I                     R                      N   
RCN12032                     I                     R                      N   
RCN13501                     I                     R                      N   
RCN12873                     I                     R                      N   
RCN12710                     I                     R                      N   
RCN13368                     I                     R                      N   
RCN12729                     I                     R                      N   
RCN13528                     I                     R                      N   
RCN12087                     I                     R                      N   
RCN12033                     I                     R                      N   
RCN13585                     I                     R                      N   
RCN12865                     I                     R                      N   
RCN18593                     I                     R                      N   
RCN13407                     I                     R                      N   
RCN15343                     I                     R                      N   
RCN12796                     I                     R                      N   
RCN25190                     I                     R                      N   
RCN13333                     I                     R                      N   
RCN14642                     I                     R                      N   
RCN15274                     I                     R                      N   
RCN12862                     I                     R                      N   
RCN12855                     I                     R                      N   
RCN12480                     I                     R                      N   
RCN12025                     I                     R                      N   
RCN25946                     I                     R                      N   
RCN12789                     I                     R                      N   
RCN12854                     I                     R                      N   
RCN12484                     I                     R                      N   
RCN15359                     I                     R                      N   

         core_mutation_dhfr_164 core_mutation_crt_72 core_mutation_crt_74  \
sample                                                                      
RCN13568                      I                    C                    I   
RCN13560                      I                    C                    I   
RCN15107                      I                    C                    I   
RCN14580                      I                    C                    I   
RCN13532                      I                    C                    I   
RCN13481                      I                    C                    I   
RCN12032                      I                    C                    I   
RCN13501                      I                    C                    I   
RCN12873                      I                    C                    I   
RCN12710                      I                    C                    I   
RCN13368                      I                    C                    I   
RCN12729                      I                    C                    I   
RCN13528                      I                    C                    I   
RCN12087 

### Save file
Sumary: this file contains 34 "representative samples", each represent a distinct pattern of drug-resistance haplotypes.

In [48]:
genre_pf8_merged_seq_gt_concord_patterns_final_df.to_csv('Pf8-GenReMekong_fully_concordant_genotype_patterns_representative_samples.csv', index=True)

***

## Retrieve FASTQ files
A Python script "ENA_data_helper.py" is provided alongside this notebook and can be used to download the FASTQ files from ENA.  
The full download is time intenstive and uses a large amount of disk space and is done performed in this notbook.  

The script works with the ENA search results produced earlier, but it needs the FASTQ FTP URLs to be split into two columns.  
In addition, the ENA search results should be filtered by sample ID against the final data saved for each of the three datasets to ensure that data is downloaded only for those samples that are still in the final dataset.

Concordant phenotype data

In [93]:
df = genre_pf8_conc_s_ena_result[
    genre_pf8_conc_s_ena_result['sample'].isin(genre_pf8_conc_s_ena2_df.index)
]
df[['ftp_url_read_1', 'ftp_url_read_2']] = df['fastq_ftp'].str.split(';', expand=True)
df = df[['run_accession','sample','ftp_url_read_1','ftp_url_read_2']]
df.to_csv('ENA_download_manifest_concordant_phenotype_high_quality_samples.csv', index=False)

In [94]:
df = genre_pf8_merged_seq_gt_concord_small_ena_result[
    genre_pf8_merged_seq_gt_concord_small_ena_result['sample'].isin(genre_pf8_merged_seq_gt_concord_small_ena2_df.index)
]
df[['ftp_url_read_1', 'ftp_url_read_2']] = df['fastq_ftp'].str.split(';', expand=True)
df = df[['run_accession','sample','ftp_url_read_1','ftp_url_read_2']]
df.to_csv('ENA_download_manifest_fully_concordant_genotype_high_quality_samples.csv', index=False)

In [95]:
df = genre_pf8_merged_seq_gt_concord_small_ena_result[
    genre_pf8_merged_seq_gt_concord_small_ena_result['sample'].isin(genre_pf8_merged_seq_gt_concord_patterns_final_df.index)
]
df[['ftp_url_read_1', 'ftp_url_read_2']] = df['fastq_ftp'].str.split(';', expand=True)
df = df[['run_accession','sample','ftp_url_read_1','ftp_url_read_2']]
df.to_csv('ENA_download_manifest_fully_concordant_genotype_patterns_representative_samples.csv', index=False)

***